metrotranscom/doorway
# feat: update listing filters to strictly match on units

# Pull Request Template

## Issue Overview


## Description

The listing filters as originally implemented applied all filters on the listing level, even when filtering on unit attributes like monthly rent or number of bedrooms.  A search for "2 bedrooms for under $800/month" would return a listing that has at least one unit with two bedrooms and at least one unit that is under $800/month, even if those were completely different units.  This PR updates that filtering logic so that it applies filtering for units on each unit in a listing and only returns a listing if at least one unit matches all search criteria.  In this case, a search for "2 bedrooms for under $800/month" will only return a listing if it has at least one unit that both has two bedrooms _and_ is under $800/month.

Some fields that need to be filtered on either now or in the future are numeric values that are stored as text (ie `"100.00"` vs `100.00`), which makes comparisons other than (in)equality impossible on the database side.  With the previous filters, this was worked around by converting from text to integer either in the view query (for local listings) or into new fields (for imported listings), which were represented in the view as min/max fields that could be filtered on.  With this approach, those aggregate fields have been removed and are replaced by new properties on the unit objects using the same conversion logic as before.  For example, where a unit may have a `monthlyRent` property with a value of `"1000"`, we now add a new property to that unit named `numMonthlyRent` with a value of `1000`.  These properties can be filtered on in our queries, but are automatically removed when translated into DTO objects to be returned by the service.

The number of migrations around combined and external listings was growing a bit out of hand, so this PR condenses them down to a single migration, removing needless flip-flopping between states.  This is safe to do as there are no persistent instances of this service running whose migration state would be negatively impacted.

## How Can This Be Tested/Reviewed?

End-to-end tests have been added to cover a few common filter combinations.  You should be able to run `yarn test:e2e:local` without any additional commands to verify that tests pass.

To test manually, first run `yarn db:reseed:with-external:dev` to rebuild the database and import listings, then build and start up the service with `yarn build` and `yarn start`.  In your browser, paste the URL below:

`http://localhost:3100/listings/combined?filter[0][$comparison]=%3E=&filter[0][numBedrooms]=2&filter[1][$comparison]=%3E=&filter[1][numBathrooms]=1&filter[2][$comparison]=IN&filter[2][counties][0]=Alameda`

This query will show only listings in Alameda county that have at least one unit with two bedrooms and one bathroom.  When I last tested it, there were four results.  If you are using Firefox or some JSON prettifier extension in Chrome, you should be able to easily navigate those listings and verify that they all have one or more units with 2 bedrooms and one bathroom.

The query below will only show listings in Alameda county that have at least one unit with two bedrooms for less than or equal to $1000/month.  There was only one such listing when I tested.
`http://localhost:3100/listings/combined?filter[0][$comparison]=%3E=&filter[0][numBedrooms]=2&filter[1][$comparison]=%3E=&filter[1][numBathrooms]=1&filter[2][$comparison]=IN&filter[2][counties][0]=Alameda`

## Checklist:


## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.
